lib/node/runner/checker_ballot_transaction.go

```diff
@@ -244,14 +254,79 @@ func ValidateOp(st *storage.LevelDBBackend, source *block.BlockAccount, op trans
 			err = errors.ErrorFrozenAccountNoDeposit
 			return
 		}
-		// If it's a frozen account, everything must be withdrawn
 		if source.Linked != "" {
+			// Source account's linked address must be equal to target address
+			if source.Linked != casted.TargetAddress() {
+				err = errors.ErrorUnfreezingToInvalidLinkedAccount
+			}
+			// If it's a frozen account, everything must be withdrawn
 			var expected common.Amount
 			expected, err = source.Balance.Sub(common.BaseFee)
 			if casted.Amount != expected {
 				err = errors.ErrorFrozenAccountMustWithdrawEverything
 				return
 			}
+			// Unfreezing must be done after X period from unfreezing request
+			var saved []block.BlockOperation
```

Comment:
For performance, we should do the check in the iteration, not allocate a new array.
Also, since it's a frozen account, there can only be 1 operation associated with this account.